Add Toast notifications to sharing UI

barrymcgee opened the pull request:
## Done

Add Toast notifications to sharing UI

## QA

- Go to model
- Open sharing panel
- Add new user / change an existing users access / delete a user
- See positive action toast
- Try to add a user which already exists
- See negative (error) toast

webteam-app commented:
Demo starting at https://jaas-dashboard-1026.demos.haus

hatched commented:
QA Notes:

The background of the toast is transparent
<img width="464" alt="Screen Shot 2021-06-07 at 2 01 17 PM" src="https://user-images.githubusercontent.com/532033/121080647-e96b1e80-c798-11eb-95d6-9bac7997e2a9.png">

There is no way to dismiss the toast. This is an issue when it covers up an action element.

When removing someone it doesn't appear to animate them closed any longer.
<img width="358" alt="Screen Shot 2021-06-07 at 2 03 16 PM" src="https://user-images.githubusercontent.com/532033/121080825-2c2cf680-c799-11eb-8b95-88abb5f0ece7.png">

hatched commented on src/components/ToastCard/ToastCard.test.tsx:
```diff
@@ -0,0 +1,35 @@
+    const wrapper = mount(
+      <ToastCard type="positive" message="I am a toast message" />
+    );
+    expect(wrapper.find("[data-type='positive']").exists()).toBe(true);
```
This doesn't check for the existence of an icon.

hatched commented on src/components/ToastCard/ToastCard.tsx:
```diff
@@ -0,0 +1,32 @@
+};
+
+export default function ToastCard({ type, message }: Props) {
+  let iconName;
```
Can you add a default here.

hatched replied:
You could also use an enum (object) which would return the correct value for the class below....a lot less lines :)

hatched commented on src/components/ToastCard/ToastCard.tsx:
```diff
@@ -0,0 +1,32 @@
+  }
+  return (
+    <div className="toast-card" data-type={type}>
+      <i className={`p-icon--${iconName}`}>Success</i>
```
Without a default this would resolve to `p-icon--`

hatched commented on src/components/ToastCard/_toast-card.scss:
```diff
@@ -0,0 +1,29 @@
+@import "vanilla-framework/scss/vanilla";
```
Vanilla has toast styles already. Did those not work for some reason?

barrymcgee replied:
As discussed, Vanilla does not have toast but notification styles which are different.

barrymcgee commented:
@hatched I've added a dismiss button to the toast notifications.

For the other bug, I've added a separate issue here: https://github.com/canonical-web-and-design/jaas-dashboard/issues/1038

hatched commented on src/components/ToastCard/ToastCard.test.tsx:
```diff
@@ -0,0 +1,73 @@
+import ToastCard from "./ToastCard";
+
+describe("Toast Card", () => {
+  const t = {
```
This object is shared between all of the test instances. While unlikely in this case it's a good idea to always clone the data for each test so that you don't end up with cascading or spurious failures. In this case it's an easy fix, just use the `cloneDeep` library we have and clone it for each test.

hatched commented on src/components/ToastCard/_toast-card.scss:
```diff
@@ -0,0 +1,35 @@
+  background-color: $color-x-light;
+  border: 1px solid $color-mid-light;
+  border-radius: 5px;
+  box-shadow: 0 0 2rem 0 rgba(51, 51, 51, 0.2);
```
Do we have a variable for a drop shadows throughout the app?

barrymcgee replied:
Seems `$box-shadow` is available in Vanilla 🤙🏻
